tests-ng/DEVELOPER.md

```diff
@@ -0,0 +1,490 @@
+3. **Testing system-level behavior** that requires shell execution
+
+```python
+# Good: Direct filesystem access
+def test_os_release():
+    with open("/etc/os-release", "r") as f:
+        content = f.read()
+        assert "ID=gardenlinux" in content
+
+# Good: Plugin abstraction for system interactions
+def test_service_status(systemd: Systemd):
+    assert systemd.is_active("ssh"), "SSH service is not running"
+
+# Acceptable: Shell call when testing shell functionality
+def test_shell_command_execution(shell: ShellRunner):
+    """Test that shell commands execute correctly."""
+    from datetime import datetime
+    result = shell("date +%Y")
+    current_year = str(datetime.now().year)
+    assert result.stdout.strip() == current_year, f"Shell command execution failed, expected {current_year}"
+
+# Bad: Shell call when plugin abstraction exists
+def test_service_status(shell: ShellRunner):
+    result = shell("systemctl is-active ssh")
+    assert result.stdout.strip() == "active", "SSH service is not running"
+```
+
+## Framework Structure
+
+### How Tests, Plugins, and Handlers Connect
+
+The framework uses pytest's plugin system to automatically register fixtures:
+
+1. **Plugins** (`tests-ng/plugins/`) - Provide fixtures for system access
+2. **Handlers** (`tests-ng/handlers/`) - Provide fixtures for setup/teardown
+3. **Tests** (`tests-ng/test_*.py`) - Use fixtures via dependency injection
+
+**Registration**: All plugins are automatically registered as pytest fixtures via `conftest.py`
+
+### Plugins (`tests-ng/plugins/`)
+
+Plugins handle infrastructure concerns and system interactions:
+
+- **Purpose**: Provide clean APIs for system access (file parsing, service management, etc.)
+- **Usage**: Provide pytest fixtures that can be injected into test functions
+- **Examples**: `Systemd`, `Sshd`, `ShellRunner`, `KernelModule`
+- **Guideline**: Handle "how to access" not "what to test"
+
+```python
+# Example plugin usage - fixtures injected into test functions
+def test_systemd_unit(systemd: Systemd):
+    assert systemd.is_active("ssh"), "SSH service is not running"
+
+def test_ssh_config(sshd: Sshd):
+    config_value = sshd.get_config_section("PermitRootLogin")
+    assert config_value == "No", f"Expected 'No', got '{config_value}'"
+```
+
+### Handlers (`tests-ng/handlers/`)
+
+Handlers are pytest fixtures that manage test setup and teardown:
+
+- **Purpose**: Setup/teardown of test state (connections, services, environments)
+- **Pattern**: Yield fixtures that prepare resources and explicitly clean up after tests
+- **Usage**: Used as pytest fixtures with `yield` for cleanup
+- **Examples**: `service_ssh`, `service_containerd`
+
+**Key distinction**: Unlike regular fixtures that provide data, handlers manage stateful resources that need explicit cleanup.
+
+```python
+# Example handler usage - setup/teardown pattern
+@pytest.fixture
+def service_ssh(systemd: Systemd):
```

Comment:
Maybe this goes too far, but sometimes it may be a good thing to set up a naming convention for handlers to distinguish them from normal features?
Something like `service_*` or `*_handler()`?

Reply:
@fwilhe What is your take on enforcing/suggesting `*_handler()` naming in the "Test Organization and Naming" section? If you don't object, I would add a small section.